src/object/srv_obj_migrate.c

```diff
@@ -3796,26 +3613,118 @@ migrate_try_obj_insert(struct migrate_pool_tls *tls, uuid_t co_uuid,
+	uint32_t        tgt_in_ver;
+};
+
+static int
+ds_pool_put_ult(void *arg)
+{
+	struct ds_pool_arg *pool_arg = (struct ds_pool_arg *)arg;
+
+	if (pool_arg->pool)
+		ds_pool_put(pool_arg->pool);
+	if (pool_arg->prop)
+		daos_prop_free(pool_arg->prop);
+	return 0;
+}
+
+static int
+ds_pool_lookup_ult(void *arg)
+{
+	int                      rc;
+	uint32_t                 rebuild_ver;
+	struct ds_pool_arg      *pool_arg = (struct ds_pool_arg *)arg;
+	struct migrate_pool_tls *tls      = NULL;
+	struct pool_target      *tgts;
+
+	rc = ds_pool_lookup(pool_arg->pool_uuid, &pool_arg->pool);
+	if (rc != 0) {
+		if (rc == -DER_SHUTDOWN) {
+			D_DEBUG(DB_REBUILD, DF_UUID " pool service is stopping.\n",
+				DP_UUID(pool_arg->pool_uuid));
+			rc = 0;
+		} else {
+			D_DEBUG(DB_REBUILD, DF_UUID " pool service is not started yet. " DF_RC "\n",
+				DP_UUID(pool_arg->pool_uuid), DP_RC(rc));
+			rc = -DER_AGAIN;
+		}
+		return rc;
+	}
+
+	ds_rebuild_running_query(pool_arg->pool_uuid, -1, &rebuild_ver, NULL, NULL);
+	if (rebuild_ver == 0 || rebuild_ver != pool_arg->rebuild_ver) {
+		rc = -DER_SHUTDOWN;
+		D_GOTO(out, rc);
+	}
+
+	tls = migrate_pool_tls_lookup(pool_arg->pool_uuid, pool_arg->rebuild_ver,
+				      pool_arg->generation);
+	if (tls) {
+		migrate_pool_tls_put(tls);
+		return rc;
+	}
+
+	D_ALLOC_PTR(pool_arg->prop);
+	if (pool_arg->prop == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	rc = ds_pool_iv_prop_fetch(pool_arg->pool, pool_arg->prop);
+	if (rc)
+		D_GOTO(out, rc);
+
+	pool_arg->pool->sp_rebuilding++;
+	rc = pool_map_find_target_by_rank_idx(pool_arg->pool->sp_map, dss_self_rank(), -1, &tgts);
+	D_ASSERT(rc == dss_tgt_nr);
+	pool_arg->tgt_status = tgts[pool_arg->tgt_id].ta_comp.co_status;
+	pool_arg->tgt_in_ver = tgts[pool_arg->tgt_id].ta_comp.co_in_ver;
+	rc                   = 0;
+out:
+	return rc;
+}
+
 int
-ds_migrate_object(struct ds_pool *pool, uuid_t po_hdl, uuid_t co_hdl, uuid_t co_uuid,
-		  uint32_t version, unsigned int generation, uint64_t max_eph, uint32_t opc,
-		  daos_unit_oid_t *oids, daos_epoch_t *epochs, daos_epoch_t *punched_epochs,
-		  unsigned int *shards, uint32_t count, unsigned int tgt_idx,
-		  uint32_t new_layout_ver)
+ds_migrate_object(uuid_t pool_uuid, uuid_t po_hdl, uuid_t co_hdl, uuid_t co_uuid, uint32_t version,
+		  unsigned int generation, uint64_t max_eph, uint32_t opc, daos_unit_oid_t *oids,
+		  daos_epoch_t *epochs, daos_epoch_t *punched_epochs, unsigned int *shards,
+		  uint32_t count, unsigned int tgt_idx, uint32_t new_layout_ver)
 {
 	struct migrate_pool_tls	*tls = NULL;
 	int			i;
 	int			rc;
+	d_rank_list_t           *svc_list;
+	struct daos_prop_entry  *entry;
+	struct ds_pool_arg       arg    = {0};
+	uint32_t                 tgt_id = dss_get_module_info()->dmi_tgt_id;
+
+	uuid_copy(arg.pool_uuid, pool_uuid);
+	arg.rebuild_ver = version;
+	arg.tgt_id      = tgt_id;
+	arg.generation  = generation;
+	rc              = dss_ult_execute(ds_pool_lookup_ult, &arg, NULL, NULL, DSS_XS_SYS, 0, 0);
```

Comment:
can we somehow cache this information so we don't have to create ULT and run those IV operations every time?